graphql_compiler/schema_transformation/merge_schemas.py

```diff
@@ -0,0 +1,170 @@
+
+    Args:
+        schemas_dict: OrderedDict[str, Document], where keys are names/identifiers of schemas,
+                      and values are ASTs describing schemas. The ASTs will not be modified
+                      by this funcion
+
+    Returns:
+        MergedSchemaDescriptor, a namedtuple that contains the AST of the merged schema,
+        and the map from names of types/query type fields to the id of the schema that they
+        came from
+
+    Raises:
+        - SchemaStructureError if the schema does not have the expected form; in particular, if
+          the AST does not represent a valid schema, if any query type field does not have the
+          same name as the type that it queries, if the schema contains type extensions or
+          input object definitions, or if the schema contains mutations or subscriptions
+        - SchemaNameConflictError if there are conflicts between the names of
+          types/interfaces/enums/scalars, or conflicts between the definition of directives
+          with the same name
+    """
+    if len(schemas_dict) == 0:
+        raise ValueError(u'Expected a nonzero number of schemas to merge.')
+
+    query_type = 'RootSchemaQuery'
+    merged_schema_ast = _basic_schema_ast(query_type)  # Document
+
+    name_id_map = {}  # Dict[str, str], name of type/interface/enum/union to schema id
+    scalars = {'String', 'Int', 'Float', 'Boolean', 'ID'}  # Set[str], user defined + builtins
+    directives = {}  # Dict[str, DirectiveDefinition]
+
+    for cur_schema_id, cur_ast in six.iteritems(schemas_dict):
+        # Prevent aliasing between output and input
+        cur_ast = deepcopy(cur_ast)
+
+        # Check input schema satisfies various structural requirements
+        check_ast_schema_is_valid(cur_ast)
+
+        cur_schema = build_ast_schema(cur_ast)
+        cur_query_type = get_query_type_name(cur_schema)
+
+        # Merge cur_ast into merged_schema_ast
+        # Concatenate new scalars, new directives, and type definitions other than the query
+        # type to definitions list
+        # Raise errors for conflicting scalars, directives, or types
+        new_definitions = cur_ast.definitions  # List[Node]
+        new_query_type_fields = None  # List[FieldDefinition]
+        for new_definition in new_definitions:
+            if isinstance(new_definition, ast_types.SchemaDefinition):
+                continue
+
+            new_name = new_definition.name.value
+
+            if (
+                isinstance(new_definition, ast_types.ObjectTypeDefinition) and
+                new_name == cur_query_type
+            ):  # query type definition
+                new_query_type_fields = new_definition.fields  # List[FieldDefinition]
+
+            elif isinstance(new_definition, ast_types.DirectiveDefinition):
+                if new_name in directives:  # existing directive
+                    if new_definition == directives[new_name]:  # definitions agree
```

Comment:
Are you sure this won't lead to an error? I think this is testing equality by reference so if you parse a directive from two different schema strings then you might get an error no?

Reply:
I’m pretty sure it checks equality by value, not reference. I tested out checking equality between two node objects with identical structures before, and there’s also a test for deduplicating directives. It’d definitely be nice to check again to be sure though, if you want to confirm.

Reply:
Yeah sorry you are right. I was thinking about GraphQLDirective objects.